core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -361,7 +361,7 @@ public long getTook() {
 
         void setTranslogLocation(Translog.Location translogLocation) {
             if (freeze.get() == null) {
-                assert failure == null : "failure has to be null to set translog location";
+                assert failure == null || translogLocation == null: "failure has to be null to set translog location";
```

Comment:
I integrated master into this branch after #22626 in d71aa16ec9d0dfd0cfb40f8b54ef6f99ea8c771e.